flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TemporalRowtimeJoin.scala

```diff
@@ -0,0 +1,326 @@
+
+  /**
+    * Incremental index generator for `leftState`'s keys.
+    */
+  private var nextLeftIndex: ValueState[JLong] = _
+
+  /**
+    * Mapping from artificial row index (generated by `nextLeftIndex`) into the left side `Row`.
+    * We can not use List to accumulate Rows, because we need efficient deletes of the oldest rows.
+    *
+    * TODO: this could be OrderedMultiMap[Jlong, Row] indexed by row's timestamp, to avoid
+    * full map traversals (if we have lots of rows on the state that exceed `currentWatermark`).
+    */
+  private var leftState: MapState[JLong, Row] = _
+
+  /**
+    * Mapping from timestamp to right side `Row`.
+    *
+    * TODO: having `rightState` as an OrderedMapState would allow us to avoid sorting cost
+    * once per watermark
+    */
+  private var rightState: MapState[JLong, Row] = _
+
+  private var registeredTimer: ValueState[JLong] = _ // JLong for correct handling of default null
+
+  private var cRowWrapper: CRowWrappingCollector = _
+  private var collector: TimestampedCollector[CRow] = _
+  private var timerService: SimpleTimerService = _
+
+  private var joinFunction: FlatJoinFunction[Row, Row, Row] = _
+
+  override def open(): Unit = {
+    val clazz = compile(
+      getRuntimeContext.getUserCodeClassLoader,
+      genJoinFuncName,
+      genJoinFuncCode)
+
+    joinFunction = clazz.newInstance()
+
+    nextLeftIndex = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong](NEXT_LEFT_INDEX_STATE_NAME, BasicTypeInfo.LONG_TYPE_INFO))
+    leftState = getRuntimeContext.getMapState(
+      new MapStateDescriptor[JLong, Row](LEFT_STATE_NAME, BasicTypeInfo.LONG_TYPE_INFO, leftType))
+    rightState = getRuntimeContext.getMapState(
+      new MapStateDescriptor[JLong, Row](RIGHT_STATE_NAME, BasicTypeInfo.LONG_TYPE_INFO, rightType))
+    registeredTimer = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong](REGISTERED_TIMER_STATE_NAME, BasicTypeInfo.LONG_TYPE_INFO))
+
+    collector = new TimestampedCollector[CRow](output)
+    cRowWrapper = new CRowWrappingCollector()
+    cRowWrapper.out = collector
+    cRowWrapper.setChange(true)
+
+    val internalTimerService = getInternalTimerService(
+      TIMERS_STATE_NAME,
+      VoidNamespaceSerializer.INSTANCE,
+      this)
+
+    timerService = new SimpleTimerService(internalTimerService)
+  }
+
+  override def processElement1(element: StreamRecord[CRow]): Unit = {
+    if (!element.getValue.change) {
+      throw new TableException(
+        s"${classOf[TemporalRowtimeJoin].getSimpleName} does not support retractions on the " +
+          s"left side.")
+    }
+
+    leftState.put(getNextLeftIndex, element.getValue.row)
+    registerSmallestTimer(getLeftTime(element.getValue.row)) // Timer to emit and clean up the state
+  }
+
+  override def processElement2(element: StreamRecord[CRow]): Unit = {
+    if (!element.getValue.change) {
+      throw new TableException(
```

Comment:
This can be validated in `DataStreamJoin` ,i.e, DataStreamRetractionRules.isAccRetract(getRight).

Reply:
But there is no `UpsertSource` in the master currently, right? If so we should ticket/remember about covering this issue when upsert sources are done. Otherwise adding this check now here, would be an untestable dead code. I have already implemented this check in `translateToPlan`, but I was stuck at writing test for it, that's why I removed it.

Reply:
OK, I will keep an eye on it when I implement `UpsertSource`. 
Maybe we can ask some advice from Timo @twalthr .